File: workspace-2025-1c/Encapsulamiento/src/cerradura/CerraduraTest.java
package cerradura;

import static org.junit.jupiter.api.Assertions.*;

import org.junit.jupiter.api.Test;

class CerraduraTest {

	@Test
	void estadoInicialTest() {
		Cerradura trabex = new Cerradura(123, 3);
		assertFalse(trabex.estaAbierta());
	}

	@Test
	void primerAperturaExitosaTest() {
		Cerradura trabex = new Cerradura(123, 3);
		trabex.abrir(123);
		assertTrue(trabex.estaAbierta());
		assertEquals(1, trabex.contarAperturasExitosas());
	}
	
	@Test 
	void intentoAbrirCerraduraAbiertaTest() {
		Cerradura trabex = new Cerradura(123, 3);
		trabex.abrir(123);
		trabex.abrir(111);		
	}

	@Test
	void primerAperturaFallidaTest() {
		Cerradura trabex = new Cerradura(123, 3);
		trabex.abrir(122);
		assertFalse(trabex.estaAbierta());
		assertEquals(0, trabex.contarAperturasExitosas());
		assertEquals(1, trabex.contarAperturasFallidas());
	}

	@Test
	void dosAperturasFallidasYLuegoAciertoTest() {
		Cerradura trabex = new Cerradura(123, 3);
		trabex.abrir(122);
		assertFalse(trabex.estaAbierta());
		trabex.abrir(132);
		assertFalse(trabex.estaAbierta());
		trabex.abrir(123);
		assertTrue(trabex.estaAbierta());
		assertEquals(1, trabex.contarAperturasExitosas());
		assertEquals(2, trabex.contarAperturasFallidas());
	}

	@Test
	void tresAperturasFallidasYbloqueoTest() {
		Cerradura trabex = new Cerradura(123, 3);
		trabex.abrir(122);
		assertFalse(trabex.estaAbierta());
		trabex.abrir(132);
		assertFalse(trabex.estaAbierta());
		trabex.abrir(223);
		assertTrue(trabex.fueBloqueada());
		assertEquals(0, trabex.contarAperturasExitosas());
		assertEquals(3, trabex.contarAperturasFallidas());
	}

	@Test
	void intentoAbrirPeroFueBloqueadaTest() {
		Cerradura trabex = new Cerradura(123, 3);
		trabex.abrir(122);
		assertFalse(trabex.estaAbierta());
		trabex.abrir(132);
		assertFalse(trabex.estaAbierta());
		trabex.abrir(223);
		assertTrue(trabex.fueBloqueada());
		assertThrows(Error.class, () -> trabex.abrir(123));

	}
	@Test
	void cienAperturasExitosasTest() {
		Cerradura trabex = new Cerradura(123, 3);
		for(int i=0; i<100; i++) {
			trabex.abrir(123);
			trabex.cerrar();
			
		}
		assertEquals(100, trabex.contarAperturasExitosas());
	}

}

File: 03-encapsulamiento/cerradura/cerradura_test.py
import unittest
from cerradura import Cerradura


class TestCerradura(unittest.TestCase):
    def test_crear_cerradura(self):
        trabex = Cerradura(123456, 3)
        self.assertTrue(trabex.esta_abierta())
        self.assertFalse(trabex.fue_bloqueada())
        self.assertEqual(0, trabex.contar_aperturas_exitosas())
        self.assertEqual(0, trabex.contar_aperturas_fallidas())

    def test_cerrada_apertura_exitosa(self):
        trabex = Cerradura(123456, 3)
        trabex.cerrar()
        self.assertTrue(trabex.esta_cerrada())
        self.assertTrue(trabex.abrir(123456))
        self.assertTrue(trabex.esta_abierta())
        self.assertEqual(1, trabex.contar_aperturas_exitosas())
        self.assertEqual(0, trabex.contar_aperturas_fallidas())

    def test_cerrada_apertura_fallida(self):
        trabex = Cerradura(123456, 3)
        trabex.cerrar()
        self.assertTrue(trabex.esta_cerrada())
        self.assertFalse(trabex.abrir(123356))
        self.assertFalse(trabex.esta_abierta())
        self.assertEqual(0, trabex.contar_aperturas_exitosas())
        self.assertEqual(1, trabex.contar_aperturas_fallidas())

    def test_cerrada_apertura_fallida_sin_bloquear(self):
        trabex = Cerradura(123456, 3)
        trabex.cerrar()
        self.assertTrue(trabex.esta_cerrada())
        trabex.abrir(123356)
        trabex.abrir(223356)
        self.assertFalse(trabex.esta_abierta())
        self.assertEqual(0, trabex.contar_aperturas_exitosas())
        self.assertEqual(2, trabex.contar_aperturas_fallidas())
        trabex.abrir(123456)
        self.assertTrue(trabex.esta_abierta())
        self.assertEqual(1, trabex.contar_aperturas_exitosas())

    def test_cerrada_apertura_fallida_con_bloqueo(self):
        trabex = Cerradura(123456, 3)
        trabex.cerrar()
        self.assertTrue(trabex.esta_cerrada())
        trabex.abrir(123356)
        trabex.abrir(123356)
        trabex.abrir(223356)
        self.assertFalse(trabex.esta_abierta())
        self.assertEqual(0, trabex.contar_aperturas_exitosas())
        self.assertEqual(3, trabex.contar_aperturas_fallidas())
        self.assertTrue(trabex.fue_bloqueada)
        with self.assertRaises(RuntimeWarning):
            trabex.abrir(123456)
        with self.assertRaises(RuntimeWarning):
            trabex.abrir(23456)


if __name__ == "__main__":
    unittest.main()
